driver-core/src/main/com/mongodb/internal/connection/AbstractMultiServerCluster.java

```diff
@@ -247,64 +246,78 @@ private boolean handleReplicaSetMemberChanged(final ServerDescription newDescrip
-                    return false;
-                }
+        if (!newDescription.isPrimary()) {
+            return true;
+        }
 
-                if (!electionId.equals(maxElectionId)) {
-                    if (LOGGER.isInfoEnabled()) {
-                        LOGGER.info(format("Setting max election id to %s from replica set primary %s", electionId,
-                                newDescription.getAddress()));
-                    }
-                    maxElectionId = electionId;
-                }
-            }
+        if (isStalePrimary(newDescription)) {
+            invalidatePotentialPrimary(newDescription);
+            return false;
+        }
 
-            if (setVersion != null
-                    && (maxSetVersion == null || setVersion.compareTo(maxSetVersion) > 0)) {
-                if (LOGGER.isInfoEnabled()) {
-                    LOGGER.info(format("Setting max set version to %d from replica set primary %s", setVersion,
-                            newDescription.getAddress()));
-                }
-                maxSetVersion = setVersion;
-            }
+        maxElectionId = nullSafeMax(newDescription.getElectionId(), maxElectionId);
+        maxSetVersion = nullSafeMax(newDescription.getSetVersion(), maxSetVersion);
 
-            if (isNotAlreadyPrimary(newDescription.getAddress())) {
-                LOGGER.info(format("Discovered replica set primary %s", newDescription.getAddress()));
-            }
-            invalidateOldPrimaries(newDescription.getAddress());
+        invalidateOldPrimaries(newDescription.getAddress());
+
+        if (isNotAlreadyPrimary(newDescription.getAddress())) {
+            LOGGER.info(format("Discovered replica set primary %s with max election id %s and max set version %d",
+                    newDescription.getAddress(), newDescription.getElectionId(), newDescription.getSetVersion()));
         }
+
         return true;
     }
 
-    private boolean isStalePrimary(final ServerDescription newDescription) {
-        if (maxSetVersion == null || maxElectionId == null) {
-            return false;
+    private boolean isStalePrimary(final ServerDescription description) {
+        ObjectId electionId = description.getElectionId();
+        Integer setVersion = description.getSetVersion();
+        if (description.getMaxWireVersion() >= SIX_DOT_ZERO_WIRE_VERSION) {
+            return nullSafeCompareTo(electionId, maxElectionId) <= 0
+                    && (nullSafeCompareTo(electionId, maxElectionId) != 0 || nullSafeCompareTo(setVersion, maxSetVersion) < 0);
+        } else {
+            return setVersion != null && description.getElectionId() != null
+                    && (nullSafeCompareTo(setVersion, maxSetVersion) < 0
+                    || nullSafeCompareTo(setVersion, maxSetVersion) == 0
+                    && nullSafeCompareTo(description.getElectionId(), maxElectionId) < 0);
         }
```

Comment:
While
```java
nullSafeCompareTo(electionId, maxElectionId) <= 0 && (nullSafeCompareTo(electionId, maxElectionId) != 0 || nullSafeCompareTo(setVersion, maxSetVersion) < 0)
```
is trivially a correct negation of
```java
nullSafeCompareTo(electionId, maxElectionId) > 0 || (nullSafeCompareTo(electionId, maxElectionId) == 0 && nullSafeCompareTo(setVersion, maxSetVersion) >= 0)
```
my brain is absolutely unable (and I tried hard, I even drew truth tables) to make any sense of the former, while the latter is perfectly clear.

Can we please rewrite this condition as follows?
```java
nullSafeCompareTo(electionId, maxElectionId) < 0 || (nullSafeCompareTo(electionId, maxElectionId) == 0 && nullSafeCompareTo(setVersion, maxSetVersion) < 0)
```

Not only this way it is clear what's going on, but also it matches the structure of the pre 6.0 condition.

Reply:
Much better.  Done.